---
title: "Figures 7 and S7"
author: "MW"
date: "2024-02-02"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r setup, echo=FALSE}
library(tidyverse)
library(ape)
library(ggtree)
```

These figures are based on sanger sequencing data from *Apilactobacillus kunkeei* colonies we isolated from collected honey bees. The input data is the partitioned analysis of multi-gene alignments. 

Set colors for the figures. 
```{r aesthetics}
site_colors = c("Oshine 1" = "#f9b93a", "Nishi-honjo" = "#c7ee34", "Ichiigawa" = "#4662d7", 
                 "Kiyokawa 1" = "#18e4b6", "Kamihaya" = "#37aaf9", 
                 "Kiyokawa 2" = "#182C61", "Hirono" = "#3B3B98", "Nagano" = "#73fe5e", 
                 "Oshine 2" = "#f56b19", "Yamauchi" = "#790403", 
                 "Shimanose" = "#ca2a04", "Higashi-honjo" = "#30123b", "Sweden" = "grey")
```

Load the tree file and root the tree. 
```{r aktree}
ak_tree <- ape::read.tree("Supplementary_data_3.nex.treefile")
ak_tree_rooted <- root.phylo(ak_tree, outgroup = "Fhon13_OX335302.1")
```

Outliers were detected using TreeShrink. We dropped these tips from further analysis. 
```{r drop_tips}
ak_tree <- drop.tip(ak_tree_rooted, c("S_11", "S_48", "S_27", "S_29", "W_32-1",  "S_5", "S_20",  "S_8", "W_149-1"))
```

Load sample metadata. 
```{r ak_metadata}
metadata <- read_csv("Supplementary_data_4.csv")

# Rename Fhon13 to its full name.
metadata[which(grepl("Fhon13", metadata$sample)), "sample"] = "Fhon13_OX335302.1"
```

Quick function to add metadata to tree, wrangle SH-aLRT and bootstrapping values for visualization, and identifiying clades to highlight. 
```{r tree_fun}
tree_fun <- function(tree, ...) {
  p <- ggtree(tree, ...) %<+% metadata
  p$data <- p$data %>% 
  rowwise() %>% 
  mutate(shalrt = ifelse(isTip == F, round(as.double(str_split_i(label, "/", 1)), digits = 0), F), 
         abayes = ifelse(isTip == F, round(as.double(str_split_i(label, "/", 2)), digits = 0), F), 
         ufbs = ifelse(isTip == F, round(as.double(str_split_i(label, "/", 3)), digits = 0), F),
         hl = ifelse(node %in% c(8:13), "C_clade", ifelse(node %in% c(14:20, 1:5), "M_clade", "non_spec")))
  return(p)
}
```

This code makes Figure 7. We removed the root for ease of visualization and visualized the SH-alrt/ultra-fast bootstrap support values at the nodes.
```{r fig7}
fig7tree <- tree_fun(drop.tip(ak_tree, "Fhon13_OX335302.1"))
fig7_plot = fig7tree + geom_tiplab(aes(color = factor(hl), label = tag), align = F, linesize = 0.25, offset = 0.0003) +
    geom_text(data = subset(fig7tree$data, !isTip & shalrt != 0 & !node %in% c(44,46,47,48,49,60,65,66,77,78,79)), 
              aes(label = paste(shalrt, ufbs, sep = "/")), size = 2.5, nudge_x = -0.001, nudge_y = 0.4) + 
    geom_text(data = subset(fig7tree$data, !isTip & shalrt != 0 & node == 77), 
            aes(label = paste(shalrt, ufbs, sep = "/")), size = 2.5, nudge_x = -0.001, nudge_y = -0.4) + 
    geom_text(data = subset(fig7tree$data, !isTip & shalrt != 0 & node %in% c(46,47,48,49,60,65,66,79)), 
            aes(label = paste(shalrt, ufbs, sep = "/")), size = 2.5, nudge_x = -0.0013, nudge_y = 0.4) + 
    geom_text(data = subset(fig7tree$data, !isTip & shalrt != 0 & node %in% c(44,78)), 
            aes(label = paste(shalrt, ufbs, sep = "/")), size = 2.5, nudge_x = -0.00145, nudge_y = 0.4) +
    xlim(0, 0.05) + 
    geom_treescale(x = 0, y = -3, width = 0.01, linesize = 0.8) +
    scale_color_manual(values = c("C_clade" = "green4", "M_clade" = "darkgoldenrod", "non_spec" = "grey36")) + 
    theme(legend.position = "none", plot.margin = unit(c(0, 0, 0, 0), "inches"))
```

This code make Figure S7 panel A. 
```{r figS7.A}
figS7.Atree <- tree_fun(ak_tree, branch.length = "none")
figS7.Atree + 
  geom_tippoint(aes(fill = factor(new_site)), size = 3, shape = 23, position = position_nudge(x = 0.05)) + 
  geom_tiplab(aes(color = factor(bee_species), label = isolate), align = F, linesize = 0.25, offset = 0.15) +
  geom_text(data = subset(figS7.Atree$data, !isTip & node != 50), aes(label = label), size = 2.5, nudge_x = -0.75, nudge_y = 0.4) + 
  geom_text(data = subset(figS7.Atree$data, !isTip & node == 50), aes(label = label), size = 2.5, nudge_x = -0.75, nudge_y = -0.4) + 
  xlim(0, 20) + 
  scale_fill_viridis(name = "Sites", discrete = T, option = "H") + 
  scale_color_manual(name = "Bee species", 
                     values = c("cerana" = "green4", "mellifera" = "darkgoldenrod"), 
                     labels = c("Apis cerana", "Apis mellifera")) + 
  theme(legend.position = "none") 
```

This code makes Figure S7 panel B. 
```{r figS7.B}
figS7.Btree <- tree_fun(ak_tree)
figS7.Btree + 
  geom_tippoint(aes(fill = factor(new_site)), size = 3, shape = 23, position = position_nudge(x = 0.001)) + 
  geom_tiplab(aes(color = factor(bee_species), label = isolate), align = F, linesize = 0.25, offset = 0.005) + 
  geom_treescale(x = 0.3, y = -3) + scale_fill_viridis(name = "Sites", discrete = T, option = "H") + 
  xlim(0, 0.7) +
  scale_color_manual(name = "Bee species", 
                     values = c("cerana" = "green4", "mellifera" = "darkgoldenrod"), 
                     labels = c("Apis cerana", "Apis mellifera")) + 
  theme(legend.position = "none")
```

Jpeg figure 7 created using the following code: 
```{r fig7_jpg}
top.vp = viewport(layout = grid.layout(3, 3, widths = unit(c(26, 228), "mm"), heights = unit(c(26, 200, 26), "mm")))
plot.vp = viewport(layout.pos.col = 2, layout.pos.row = 2, name = "plot")
splot = vpTree(top.vp, vpList(plot.vp))

species_legend = Legend(at = c("Apis cerana", "Apis mellifera"), title = "Bee species", title_position = c("topcenter"), title_gp = gpar(fontsize = 13, fontface = "bold"), labels_gp = gpar(fontface = "italic", fontsize = 12, col = c("green4", "darkgoldenrod")))

jpeg(filename = "Fig7.jpg", width = 254, height = 252, units = "mm", pointsize = 16, res = 400)
pushViewport(splot)
seekViewport("plot")
grid.draw(ggplotGrob(fig7_plot))
draw(species_legend, x = unit(0.7, "npc"), y = unit(0.5, "npc"), just = c("right", "bottom"))
popViewport()
dev.off()
```













